lib/maths/unittest/CDataFrameUtilsTest.cc

```diff
@@ -0,0 +1,109 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+#include "CDataFrameUtilsTest.h"
+
+#include <maths/CBasicStatistics.h>
+#include <maths/CDataFrameUtils.h>
+
+#include <test/CRandomNumbers.h>
+
+#include <boost/filesystem.hpp>
+
+#include <vector>
+
+using namespace ml;
+
+void CDataFrameUtilsTest::testStandardizeColumns() {
+
+    using TDoubleVec = std::vector<double>;
+    using TDoubleVecVec = std::vector<TDoubleVec>;
+    using TMeanVarAccumulatorVec =
+        std::vector<maths::CBasicStatistics::SSampleMeanVar<double>::TAccumulator>;
+    using TFactoryFunc = std::function<std::unique_ptr<core::CDataFrame>()>;
+
+    test::CRandomNumbers rng;
+
+    std::size_t rows{2000};
+    std::size_t cols{4};
+    std::size_t capacity{500};
+
+    TDoubleVecVec values(4);
+    TMeanVarAccumulatorVec moments(4);
+    {
+        std::size_t i = 0;
+        for (auto a : {-10.0, 0.0}) {
+            for (auto b : {5.0, 30.0}) {
+                rng.generateUniformSamples(a, b, rows, values[i++]);
+            }
+        }
+        for (i = 0; i < cols; ++i) {
+            moments[i].add(values[i]);
+        }
+    }
+
+    TFactoryFunc makeOnDisk{[=] {
+        return core::makeDiskStorageDataFrame(
+                   boost::filesystem::current_path().string(), cols, rows, capacity)
+            .first;
+    }};
+    TFactoryFunc makeMainMemory{
+        [=] { return core::makeMainStorageDataFrame(cols, capacity).first; }};
+
+    core::stopDefaultAsyncExecutor();
+
+    for (auto threads : {1, 4}) {
+
+        for (const auto& factory : {makeOnDisk, makeMainMemory}) {
+
+            auto frame = factory();
+
+            for (std::size_t i = 0; i < rows; ++i) {
+                frame->writeRow([&values, i, cols](core::CDataFrame::TFloatVecItr column,
+                                                   std::int32_t&) {
+                    for (std::size_t j = 0; j < cols; ++j, ++column) {
+                        *column = values[j][i];
+                    }
+                });
+            }
+            frame->finishWritingRows();
+
+            CPPUNIT_ASSERT(maths::CDataFrameUtils::standardizeColumns(threads, *frame));
+
+            bool passed{true};
+            frame->readRows(1, [&](core::CDataFrame::TRowItr beginRows,
+                                   core::CDataFrame::TRowItr endRows) {
+                for (auto row = beginRows; row != endRows; ++row) {
+                    for (std::size_t j = 0; j < row->numberColumns(); ++j) {
+                        double mean{maths::CBasicStatistics::mean(moments[j])};
```

Comment:
I wonder if we should also have a test with few, simple numbers so we can assert on the result directly instead of using the same code the function we're testing is using.

Reply:
I see this as more testing that the operations applied directly to the values we expect to be in the data frame are equivalent. The test code is not going via the data frame so all code related to interacting with the data frame is tested, including threading and partitioning the values up to support this.

The question is then do we also want to retest `maths::CBasicStatistics` here, since (amongst other things) it is unit tested with basic cases? I'm not sure this is worthwhile.

However, there is also a couple of invariants which could be checked, i.e. that the mean should be zero and the variance should be one, when computed on the standardised column values. These could be included as another source of "independent" asserts.

Reply:
[f94cfa7](https://github.com/elastic/ml-cpp/pull/371/commits/f94cfa7c5ab563740d07e8472deea1f687a457d5)